csrc/scheduler/multi_matmul.cpp

```diff
@@ -0,0 +1,426 @@
+        // There can be multiple uses for example if we have A @ B1 + A @ B2
+        // then A will be cached to smem then it might be loaded into two
+        // separate register buffers, one for each mma. Instead, we will load
+        // it once into registers then re-use the register buffer for both
+        // mmas.
+        cacheAfter(smem_operands[i]);
+      }
+      NVF_ERROR(smem_operands[i]->uses().size() == 1);
+      smem_operands[i]->setMemoryType(MemoryType::Shared);
+    }
+  }
+
+  // We add two LoadStore operators to the inputs of our fusions. The first
+  // one is for a read from global memory and the second one (below) is for a
+  // cache read. As an optimizaton, we avoid adding an operator if there's an
+  // existing LoadStoreOp present. Please note that for the second LoadStore
+  // we don't propagate the allocation domain, since the scheduler sets the
+  // allocation domain in the registers.
+  void addSetsForCacheReads(
+      const std::vector<TensorView*>& tv_smems,
+      std::vector<TensorView*>& tv_rs) {
+    tv_rs.resize(tv_smems.size(), nullptr);
+    for (size_t i : c10::irange(tv_smems.size())) {
+      TensorView* tv_smem = tv_smems[i];
+      TensorView*& tv_r = tv_rs[i];
+
+      if (auto ldst = dynamic_cast<LoadStoreOp*>(tv_smem->uses().at(0))) {
+        tv_r = ldst->out()->as<TensorView>();
+        ldst->setOpType(LoadStoreOpType::LdMatrix);
+      } else {
+        tv_r = cacheAfter(
+            tv_smem,
+            LoadStoreOpType::LdMatrix,
+            CacheOp::Unspecified,
+            /*propagate_allocation_domain=*/false);
+      }
+    }
+  }
+
+  //! This calls orig->cacheAfter() and also updates the permissive graph to
+  //! reflect the new IterDomain mappings
+  TensorView* cacheAfter(
+      TensorView* orig,
+      LoadStoreOpType op_type = LoadStoreOpType::Set,
+      CacheOp cache_op = CacheOp::AllLevels,
+      bool propagate_allocation_domain = false) {
+    const std::vector<IterDomain*> orig_alloc =
+        orig->getMaybeAllocationDomain();
+
+    TensorView* c =
+        orig->cacheAfter(op_type, cache_op, propagate_allocation_domain);
+
+    if (propagate_allocation_domain) {
+      const std::vector<IterDomain*> cache_alloc =
+          c->getMaybeAllocationDomain();
+      NVF_ERROR(orig_alloc.size() == cache_alloc.size());
+      for (size_t i : c10::irange(orig_alloc.size())) {
+        ValGroup vg = graph_->toGroup(orig_alloc[i]);
+        graph_->initializeVal(cache_alloc[i], vg);
+      }
+    }
+
+    const std::vector<IterDomain*> orig_logical =
+        TensorDomain::noReductions(orig->getLogicalDomain());
+    const std::vector<IterDomain*> cache_logical = c->getLogicalDomain();
+    // in split-K we do rFactor which gives us a full = sum(partial)
+    // where partial has root domain that matches the logical domain of the
+    // original tensor. The logical domain contains Iteration transforms of the
+    // Reduction axis in the original mma output.
```

Comment:
@zasdfgbnm this code comment is relevant to your suggestion to use `tv->allIDs()` and map all the corresponding `ValGroup`s. If we have an rfactored partial reduction it can have `Iteration` transforms of a `Reduction` axis and vice versa. Something like this happens for split-K with smem epilogue since we place the smem tensor between the partial and full reduction tensors using `partial->cacheAfter()`:
```
original mma_result: T12_l[ iS194{( ceilDiv(i0, 256) )}, iS196{( ceilDiv(i4, 128) )}, iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf, iS195{256}, iS197{128}, rS199{32}rf ]
 root domain : (iS191{i0}, iS192{i4}, rS193{i2}rf)
  Split: rS193{i2}rf by factor 32 -> iS198{( ceilDiv(i2, 32) )}rf, rS199{32}rf
  Outer split: iS198{( ceilDiv(i2, 32) )}rf by factor 2 -> iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf
 logical domain : (iS191{i0}, iS192{i4}, iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf, rS199{32}rf)
 contiguity: t t t n n
  Split: iS191{i0} by factor 256 -> iS194{( ceilDiv(i0, 256) )}, iS195{256}
  Split: iS192{i4} by factor 128 -> iS196{( ceilDiv(i4, 128) )}, iS197{128}
 loop domain : (iS194{( ceilDiv(i0, 256) )}, iS196{( ceilDiv(i4, 128) )}, iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf, iS195{256}, iS197{128}, rS199{32}rf)
 mma_result after cacheAfter(): T12_l[ iS194{( ceilDiv(i0, 256) )}, iS196{( ceilDiv(i4, 128) )}, iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf, iS195{256}, iS197{128}, rS199{32}rf ]
 root domain : (iS191{i0}, iS192{i4}, rS193{i2}rf)
  Split: rS193{i2}rf by factor 32 -> iS198{( ceilDiv(i2, 32) )}rf, rS199{32}rf
  Outer split: iS198{( ceilDiv(i2, 32) )}rf by factor 2 -> iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf
 logical domain : (iS191{i0}, iS192{i4}, iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf, rS199{32}rf)
 contiguity: t t t n n
  Split: iS191{i0} by factor 256 -> iS194{( ceilDiv(i0, 256) )}, iS195{256}
  Split: iS192{i4} by factor 128 -> iS196{( ceilDiv(i4, 128) )}, iS197{128}
 loop domain : (iS194{( ceilDiv(i0, 256) )}, iS196{( ceilDiv(i4, 128) )}, iS200{2}rf, rS201{( ceilDiv(( ceilDiv(i2, 32) ), 2) )}rf, iS195{256}, iS197{128}, rS199{32}rf)
cache after cacheAfter(): T13_l[ iS212{( ceilDiv(i0, 256) )}, iS214{( ceilDiv(i4, 128) )}, iS211{2}, iS213{256}, iS215{128} ]
 logical domain : (iS209{i0}, iS210{i4}, iS211{2})
 contiguity: t t t
  Split: iS209{i0} by factor 256 -> iS212{( ceilDiv(i0, 256) )}, iS213{256}
  Split: iS210{i4} by factor 128 -> iS214{( ceilDiv(i4, 128) )}, iS215{128}
 loop domain : (iS212{( ceilDiv(i0, 256) )}, iS214{( ceilDiv(i4, 128) )}, iS211{2}, iS213{256}, iS215{128})
```
Determining the corresponding IDs to do the mapping here is tricky since any of the original reduction domains are missing in the cache tensor, and because the pre-cache tensor has root->logical transforms while the cache tensor does not.

Note that I only need this graph in order to determine the leaf and allocation dim roles, so if the other IDs are not mapped it is not really a problem for me. However, a more general way to incrementally update the IdModel's graphs during scheduling (creation of tensors with cache* and rFactor) would be generally useful it seems.

Reply:
So you will not use the same IdModel for transform propagation?

Reply:
Currently I'm not using IdModel to propagate transforms anywhere; this is because of the questions that came up around forwarding around merges with missing axes. You're right though that in the future it would be nice to be able to keep the permissive graph intact (or something like it like EXACT+concretized bcasts as we discussed at the on-site meeting recently). To do that, I think it would probably be best to actually do the mapping at the TV level, using some `IdModel` method that would update all of the graphs. Then we would just build the IdModel once, modify its Exact graph to map concretized broadcasts into their Iteration groups, then whenever we create a new TV it would be updated just like the exact graph, which I think would give us all these intermediate ID mappings in the natural way.